### ISSUE TITLE
Floating point addition is not commutative - Unexpected behavior - leads to NaN issues

### ISSUE THREAD

**Issue by Spenhouet** [tensorflow/tensorflow#25728]
**System information**
- Have I written custom code (as opposed to using a stock example script provided in TensorFlow): No
- OS Platform and Distribution (e.g., Linux Ubuntu 16.04): Windows and Linux
- TensorFlow installed from (source or binary): binary
- TensorFlow version (use command below): b'v1.12.0-rc2-3-ga6d8ffae09' 1.12.0
- Python version: Python 3.6.5
- CUDA/cuDNN version: v9.0 / 7.2.1
- GPU model and memory: GTX 1080 and also on GTX 1080 TI

**Describe the current behavior**

Swapping the position of operands gives different results:

- `1e-8 + 1.0 - tf.identity(1.0)` results in `0.0`
- `1.0 - tf.identity(1.0) + 1e-8` results in `1e-08`

**Describe the expected behavior**

Operations like addition are commutative. Swapping the position of operands doesn't change the result:

- `1e-8 + 1.0 - tf.identity(1.0)` results in `1e-08`
- `1.0 - tf.identity(1.0) + 1e-8` results in `1e-08`

**Code to reproduce the issue**

```
import tensorflow as tf
sess = tf.Session()

print(sess.run(1e-8 + 1.0 - tf.identity(1.0)))
print(sess.run(1.0 - tf.identity(1.0) + 1e-8))
```

Output:

> 0.0
> 1e-08

Expected output:

> 1e-08
> 1e-08

**Additional comment**

In combination with `tf.log` / `tf.divide` this results in `-inf` / `inf` instead of `-18.420681` / `100000000.0`
Working with further calculations, for example log loss,  this results in a sudden `NaN` error without any exploding or other indicators. This is unexpected and leads to confusion. Many people on the internet who have inexplicable `NaN` issues try a multitude of proposed solutions (use other weight init, lower learning rate, clipping, strange `tf.where` `NaN` treatment, ....) to work around this unexpected behavior. I don't think any of the before mentioned solutions is appropriate to handle this unexpected behavior and I wish for it to work as expected.

**Comment by freedomtan** [comment]
I don't think floating point addition is commutative.

**Comment by rmothukuru** [comment]
@Spenhouet,
Sorry for the delayed response. This [Stack Overflow Answer](https://stackoverflow.com/questions/24442725/is-floating-point-addition-commutative-in-c#:~:text=The%20mathematical%20operation%20%22sum%22%20is,addition%20is%20definitely%20not%20associative.) explains that **`Floating Point Addition`** in **`C++`** is not **`Commutative`**.  

So, the behavior can be considered as expected. Thanks!
